runtime/floats.c

```diff
@@ -372,51 +370,63 @@ static int caml_float_of_hex(const char * s, double * res)
   return 0;
 }
 
-CAMLprim value caml_float_of_string(value vs)
+static int caml_float_of_string_no_underscore(const char * src, double * res)
 {
-  char parse_buffer[64];
-  char * buf, * dst, * end;
-  const char *src;
-  mlsize_t len;
-  int sign;
-  double d;
-
   /* Check for hexadecimal FP constant */
-  src = String_val(vs);
-  sign = 1;
+  int sign = 1;
   if (*src == '-') { sign = -1; src++; }
   else if (*src == '+') { src++; };
   if (src[0] == '0' && (src[1] == 'x' || src[1] == 'X')) {
-    if (caml_float_of_hex(src + 2, &d) == -1)
-      caml_failwith("float_of_string");
-    return caml_copy_double(sign < 0 ? -d : d);
+    if (caml_float_of_hex(src + 2, res) == -1)
+      return -1;
+    if (sign < 0) *res = - (* res);
+    return 0;
   }
-  /* Remove '_' characters before calling strtod () */
-  len = caml_string_length(vs);
-  buf = len < sizeof(parse_buffer) ? parse_buffer : caml_stat_alloc(len + 1);
-  src = String_val(vs);
-  dst = buf;
-  while (len--) {
-    char c = *src++;
-    if (c != '_') *dst++ = c;
-  }
-  *dst = 0;
-  if (dst == buf) goto error;
+  else {
+    char * end;
 #if defined(HAS_STRTOD_L) && defined(HAS_LOCALE)
-  d = strtod_l((const char *) buf, &end, caml_locale);
+    *res = strtod_l(src, &end, caml_locale);
```

Comment:
This diff is not correct, `src` was potentially moved already and no longer contains the sign.

Reply:
fixed